src/ray/core_worker/lib/java/jni_utils.h

```diff
@@ -566,13 +568,26 @@ inline jobject NativeRayObjectToJavaNativeRayObject(
   if (!rayObject) {
     return nullptr;
   }
-  auto java_data = NativeBufferToJavaByteArray(env, rayObject->GetData());
+  auto metadata_ptr = rayObject->GetMetadata();
+  auto metadata_str = std::string(reinterpret_cast<const char*>(metadata_ptr->Data()));
   auto java_metadata = NativeBufferToJavaByteArray(env, rayObject->GetMetadata());
-  auto java_obj = env->NewObject(java_native_ray_object_class,
+  jobject java_obj;
+  if (metadata_str.length() >=5 && metadata_str.substr(0, 5).compare("ARROW") == 0) {
+    auto data_ptr = rayObject->GetData();
+    auto java_data = env->NewDirectByteBuffer(
```

Comment:
Better combine this with the existing `ByteBuffer` support. See https://github.com/ray-project/ray/blob/117219557190101d5a7cf776843e602ec5d06e74/src/ray/core_worker/lib/java/io_ray_runtime_RayNativeRuntime.cc#L59-L65 and https://github.com/ray-project/ray/blob/117219557190101d5a7cf776843e602ec5d06e74/java/runtime/src/main/java/io/ray/runtime/task/ArgumentsBuilder.java#L86-L88.

Though the existing `ByteBuffer` support only applies to task arguments, not `Ray.get()`. We may need to support `Ray.get()` as well.

Reply:
When I use direct byte buffer, I meant to eliminate a copy for Arrow reading, not directly using it as ByteBuffer. So I need metadata in the NativeRayObject. I guess these args can also use NativeRayObject.buffer, but that's beyond this pr 's scope.